// @owlbear-rodeo/sdk/lib/api/scene/SceneApi.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import SceneFogApi from "./SceneFogApi";
import SceneGridApi from "./SceneGridApi";
import SceneHistoryApi from "./SceneHistoryApi";
import SceneItemsApi from "./SceneItemsApi";
import SceneLocalApi from "./SceneLocalApi";
class SceneApi {
    constructor(messageBus) {
        this.messageBus = messageBus;
        this.grid = new SceneGridApi(messageBus);
        this.fog = new SceneFogApi(messageBus);
        this.history = new SceneHistoryApi(messageBus);
        this.items = new SceneItemsApi(messageBus);
        this.local = new SceneLocalApi(messageBus);
    }
    isReady() {
        return __awaiter(this, void 0, void 0, function* () {
            const { ready } = yield this.messageBus.sendAsync("OBR_SCENE_IS_READY", {});
            return ready;
        });
    }
    onReadyChange(callback) {
        const handleChange = (data) => {
            callback(data.ready);
        };
        this.messageBus.send("OBR_SCENE_READY_SUBSCRIBE", {});
        this.messageBus.on("OBR_SCENE_EVENT_READY_CHANGE", handleChange);
        return () => {
            this.messageBus.send("OBR_SCENE_READY_UNSUBSCRIBE", {});
            this.messageBus.off("OBR_SCENE_EVENT_READY_CHANGE", handleChange);
        };
    }
    getMetadata() {
        return __awaiter(this, void 0, void 0, function* () {
            const { metadata } = yield this.messageBus.sendAsync("OBR_SCENE_GET_METADATA", {});
            return metadata;
        });
    }
    setMetadata(update) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.messageBus.sendAsync("OBR_SCENE_SET_METADATA", { update });
        });
    }
    onMetadataChange(callback) {
        const handleChange = (data) => {
            callback(data.metadata);
        };
        this.messageBus.send("OBR_SCENE_METADATA_SUBSCRIBE", {});
        this.messageBus.on("OBR_SCENE_METADATA_EVENT_CHANGE", handleChange);
        return () => {
            this.messageBus.send("OBR_SCENE_METADATA_UNSUBSCRIBE", {});
            this.messageBus.off("OBR_SCENE_METADATA_EVENT_CHANGE", handleChange);
        };
    }
}
export default SceneApi;
