// @owlbear-rodeo/sdk/lib/api/scene/SceneHistoryApi.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
class SceneHistoryApi {
    constructor(messageBus) {
        this.messageBus = messageBus;
    }
    undo() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.messageBus.sendAsync("OBR_SCENE_HISTORY_UNDO", {});
        });
    }
    redo() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.messageBus.sendAsync("OBR_SCENE_HISTORY_REDO", {});
        });
    }
    canUndo() {
        return __awaiter(this, void 0, void 0, function* () {
            const { canUndo } = yield this.messageBus.sendAsync("OBR_SCENE_HISTORY_CAN_UNDO", {});
            return canUndo;
        });
    }
    canRedo() {
        return __awaiter(this, void 0, void 0, function* () {
            const { canRedo } = yield this.messageBus.sendAsync("OBR_SCENE_HISTORY_CAN_REDO", {});
            return canRedo;
        });
    }
}
export default SceneHistoryApi;
