// @owlbear-rodeo/sdk/lib/api/RoomApi.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
class RoomApi {
    constructor(messageBus) {
        this.messageBus = messageBus;
    }
    get id() {
        return this.messageBus.roomId;
    }
    getPermissions() {
        return __awaiter(this, void 0, void 0, function* () {
            const { permissions } = yield this.messageBus.sendAsync("OBR_ROOM_GET_PERMISSIONS", {});
            return permissions;
        });
    }
    getMetadata() {
        return __awaiter(this, void 0, void 0, function* () {
            const { metadata } = yield this.messageBus.sendAsync("OBR_ROOM_GET_METADATA", {});
            return metadata;
        });
    }
    setMetadata(update) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.messageBus.sendAsync("OBR_ROOM_SET_METADATA", { update });
        });
    }
    onMetadataChange(callback) {
        const handleChange = (data) => {
            callback(data.metadata);
        };
        this.messageBus.send("OBR_ROOM_METADATA_SUBSCRIBE", {});
        this.messageBus.on("OBR_ROOM_METADATA_EVENT_CHANGE", handleChange);
        return () => {
            this.messageBus.send("OBR_METADATA_ROOM_UNSUBSCRIBE", {});
            this.messageBus.off("OBR_ROOM_METADATA_EVENT_CHANGE", handleChange);
        };
    }
    onPermissionsChange(callback) {
        const handleChange = (data) => {
            callback(data.permissions);
        };
        this.messageBus.send("OBR_ROOM_PERMISSIONS_SUBSCRIBE", {});
        this.messageBus.on("OBR_ROOM_PERMISSIONS_EVENT_CHANGE", handleChange);
        return () => {
            this.messageBus.send("OBR_PERMISSIONS_ROOM_UNSUBSCRIBE", {});
            this.messageBus.off("OBR_ROOM_PERMISSIONS_EVENT_CHANGE", handleChange);
        };
    }
}
export default RoomApi;
