// @owlbear-rodeo/sdk/lib/builders/CurveBuilder.js
import { GenericItemBuilder } from "./GenericItemBuilder";
export class CurveBuilder extends GenericItemBuilder {
    constructor(player) {
        super(player);
        this._points = [];
        this._style = {
            fillColor: "black",
            fillOpacity: 1,
            strokeColor: "white",
            strokeOpacity: 1,
            strokeWidth: 5,
            strokeDash: [],
            tension: 0.5,
        };
        this._item.name = "Curve";
        this._item.layer = "DRAWING";
    }
    points(points) {
        this._points = points;
        return this.self();
    }
    style(style) {
        this._style = style;
        return this.self();
    }
    fillColor(fillColor) {
        this._style.fillColor = fillColor;
        return this.self();
    }
    fillOpacity(fillOpacity) {
        this._style.fillOpacity = fillOpacity;
        return this.self();
    }
    strokeColor(strokeColor) {
        this._style.strokeColor = strokeColor;
        return this.self();
    }
    strokeOpacity(strokeOpacity) {
        this._style.strokeOpacity = strokeOpacity;
        return this.self();
    }
    strokeWidth(strokeWidth) {
        this._style.strokeWidth = strokeWidth;
        return this.self();
    }
    strokeDash(strokeDash) {
        this._style.strokeDash = strokeDash;
        return this.self();
    }
    tension(tension) {
        this._style.tension = tension;
        return this.self();
    }
    closed(closed) {
        this._style.closed = closed;
        return this.self();
    }
    build() {
        return Object.assign(Object.assign({}, this._item), { type: "CURVE", points: this._points, style: this._style });
    }
}
