// @owlbear-rodeo/sdk/lib/builders/GenericItemBuilder.js
import { v4 as uuid } from "uuid";
export class GenericItemBuilder {
    constructor(player) {
        this._item = {
            createdUserId: player.id,
            id: uuid(),
            name: "Item",
            zIndex: Date.now(),
            lastModified: new Date().toISOString(),
            lastModifiedUserId: player.id,
            locked: false,
            metadata: {},
            position: { x: 0, y: 0 },
            rotation: 0,
            scale: { x: 1, y: 1 },
            type: "ITEM",
            visible: true,
            layer: "POPOVER",
        };
    }
    createdUserId(createdUserId) {
        this._item.createdUserId = createdUserId;
        return this.self();
    }
    id(id) {
        this._item.id = id;
        return this.self();
    }
    name(name) {
        this._item.name = name;
        return this.self();
    }
    lastModified(lastModified) {
        this._item.lastModified = lastModified;
        return this.self();
    }
    zIndex(zIndex) {
        this._item.zIndex = zIndex;
        return this.self();
    }
    lastModifiedUserId(lastModifiedUserId) {
        this._item.lastModifiedUserId = lastModifiedUserId;
        return this.self();
    }
    locked(locked) {
        this._item.locked = locked;
        return this.self();
    }
    metadata(metadata) {
        this._item.metadata = metadata;
        return this.self();
    }
    position(position) {
        this._item.position = position;
        return this.self();
    }
    rotation(rotation) {
        this._item.rotation = rotation;
        return this.self();
    }
    scale(scale) {
        this._item.scale = scale;
        return this.self();
    }
    visible(visible) {
        this._item.visible = visible;
        return this.self();
    }
    attachedTo(attachedTo) {
        this._item.attachedTo = attachedTo;
        return this.self();
    }
    layer(layer) {
        this._item.layer = layer;
        return this.self();
    }
    disableHit(disable) {
        this._item.disableHit = disable;
        return this.self();
    }
    disableAutoZIndex(disable) {
        this._item.disableAutoZIndex = disable;
        return this.self();
    }
    self() {
        // @ts-ignore
        return this;
    }
}
