// @owlbear-rodeo/sdk/lib/builders/LineBuilder.js
import { GenericItemBuilder } from "./GenericItemBuilder";
export class LineBuilder extends GenericItemBuilder {
    constructor(player) {
        super(player);
        this._style = {
            strokeColor: "black",
            strokeOpacity: 1,
            strokeWidth: 5,
            strokeDash: [],
        };
        this._startPosition = { x: 0, y: 0 };
        this._endPosition = { x: 0, y: 0 };
        this._item.layer = "DRAWING";
        this._item.name = "Line";
    }
    style(style) {
        this._style = style;
        return this.self();
    }
    strokeColor(strokeColor) {
        this._style.strokeColor = strokeColor;
        return this.self();
    }
    strokeOpacity(strokeOpacity) {
        this._style.strokeOpacity = strokeOpacity;
        return this.self();
    }
    strokeWidth(strokeWidth) {
        this._style.strokeWidth = strokeWidth;
        return this.self();
    }
    strokeDash(strokeDash) {
        this._style.strokeDash = strokeDash;
        return this.self();
    }
    startPosition(startPosition) {
        this._startPosition = startPosition;
        return this.self();
    }
    endPosition(endPosition) {
        this._endPosition = endPosition;
        return this.self();
    }
    build() {
        return Object.assign(Object.assign({}, this._item), { type: "LINE", startPosition: this._startPosition, endPosition: this._endPosition, style: this._style });
    }
}
