// @owlbear-rodeo/sdk/lib/builders/PathBuilder.js
import { GenericItemBuilder } from "./GenericItemBuilder";
export class PathBuilder extends GenericItemBuilder {
    constructor(player) {
        super(player);
        this._commands = [];
        this._fillRule = "nonzero";
        this._style = {
            fillColor: "black",
            fillOpacity: 1,
            strokeColor: "white",
            strokeOpacity: 1,
            strokeWidth: 5,
            strokeDash: [],
        };
        this._item.name = "Path";
        this._item.layer = "DRAWING";
    }
    commands(commands) {
        this._commands = commands;
        return this.self();
    }
    fillRule(fillRule) {
        this._fillRule = fillRule;
        return this.self();
    }
    style(style) {
        this._style = style;
        return this.self();
    }
    fillColor(fillColor) {
        this._style.fillColor = fillColor;
        return this.self();
    }
    fillOpacity(fillOpacity) {
        this._style.fillOpacity = fillOpacity;
        return this.self();
    }
    strokeColor(strokeColor) {
        this._style.strokeColor = strokeColor;
        return this.self();
    }
    strokeOpacity(strokeOpacity) {
        this._style.strokeOpacity = strokeOpacity;
        return this.self();
    }
    strokeWidth(strokeWidth) {
        this._style.strokeWidth = strokeWidth;
        return this.self();
    }
    strokeDash(strokeDash) {
        this._style.strokeDash = strokeDash;
        return this.self();
    }
    build() {
        return Object.assign(Object.assign({}, this._item), { type: "PATH", commands: this._commands, fillRule: this._fillRule, style: this._style });
    }
}
