// @owlbear-rodeo/sdk/lib/builders/RulerBuilder.js
import { GenericItemBuilder } from "./GenericItemBuilder";
export class RulerBuilder extends GenericItemBuilder {
    constructor(player) {
        super(player);
        this._startPosition = { x: 0, y: 0 };
        this._endPosition = { x: 0, y: 0 };
        this._measurement = "";
        this._item.layer = "RULER";
        this._item.name = "Ruler";
        this._style = {
            variant: "FILLED",
        };
    }
    startPosition(startPosition) {
        this._startPosition = startPosition;
        return this.self();
    }
    endPosition(endPosition) {
        this._endPosition = endPosition;
        return this.self();
    }
    measurement(measurement) {
        this._measurement = measurement;
        return this.self();
    }
    variant(variant) {
        this._style.variant = variant;
        return this.self();
    }
    build() {
        return Object.assign(Object.assign({}, this._item), { type: "RULER", startPosition: this._startPosition, endPosition: this._endPosition, measurement: this._measurement, style: this._style });
    }
}
