// @owlbear-rodeo/sdk/lib/builders/TextBuilder.js
import { GenericItemBuilder } from "./GenericItemBuilder";
export class TextBuilder extends GenericItemBuilder {
    constructor(player) {
        super(player);
        this._text = {
            richText: [
                {
                    type: "paragraph",
                    children: [{ text: "" }],
                },
            ],
            plainText: "",
            style: {
                padding: 0,
                fontFamily: "Roboto",
                fontSize: 16,
                fontWeight: 400,
                textAlign: "LEFT",
                textAlignVertical: "TOP",
                fillColor: "white",
                fillOpacity: 1,
                strokeColor: "white",
                strokeOpacity: 1,
                strokeWidth: 0,
                lineHeight: 1.5,
            },
            type: "RICH",
            width: "AUTO",
            height: "AUTO",
        };
        this._item.layer = "TEXT";
        this._item.name = "Text";
    }
    text(text) {
        this._text = text;
        return this.self();
    }
    width(width) {
        this._text.width = width;
        return this.self();
    }
    height(height) {
        this._text.height = height;
        return this.self();
    }
    richText(richText) {
        this._text.richText = richText;
        return this.self();
    }
    plainText(plainText) {
        this._text.plainText = plainText;
        return this.self();
    }
    textType(textType) {
        this._text.type = textType;
        return this.self();
    }
    padding(padding) {
        this._text.style.padding = padding;
        return this.self();
    }
    fontFamily(fontFamily) {
        this._text.style.fontFamily = fontFamily;
        return this.self();
    }
    fontSize(fontSize) {
        this._text.style.fontSize = fontSize;
        return this.self();
    }
    fontWeight(fontWeight) {
        this._text.style.fontWeight = fontWeight;
        return this.self();
    }
    textAlign(textAlign) {
        this._text.style.textAlign = textAlign;
        return this.self();
    }
    textAlignVertical(textAlignVertical) {
        this._text.style.textAlignVertical = textAlignVertical;
        return this.self();
    }
    fillColor(fillColor) {
        this._text.style.fillColor = fillColor;
        return this.self();
    }
    fillOpacity(fillOpacity) {
        this._text.style.fillOpacity = fillOpacity;
        return this.self();
    }
    strokeColor(strokeColor) {
        this._text.style.strokeColor = strokeColor;
        return this.self();
    }
    strokeOpacity(strokeOpacity) {
        this._text.style.strokeOpacity = strokeOpacity;
        return this.self();
    }
    strokeWidth(strokeWidth) {
        this._text.style.strokeWidth = strokeWidth;
        return this.self();
    }
    lineHeight(lineHeight) {
        this._text.style.lineHeight = lineHeight;
        return this.self();
    }
    build() {
        return Object.assign(Object.assign({}, this._item), { type: "TEXT", text: this._text });
    }
}
